c/tskit/trees.c

```diff
@@ -2103,22 +2126,635 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
+        if (ret != 0) {
+            goto out;
+        }
+
+        tree_index++;
+        t_left = t_right;
+    }
+out:
+    if (parent != NULL) {
+        free(parent);
+    }
+    if (right_child != NULL) {
+        free(right_child);
+    }
+    if (left_sib != NULL) {
+        free(left_sib);
+    }
+    if (right_sib != NULL) {
+        free(right_sib);
+    }
+    return ret;
+}
+
+static int
+tsk_treeseq_two_site_general_stat(const tsk_treeseq_t *self, tsk_size_t state_dim,
+    const double *sample_weights, tsk_size_t result_dim, general_stat_func_t *f,
+    sample_count_stat_params_t *f_params, norm_func_t *norm_f,
+    const double *TSK_UNUSED(left_window), const double *TSK_UNUSED(right_window),
+    tsk_flags_t options, tsk_size_t *result_size, double **result)
+{
+    int ret = 0;
+    const tsk_size_t num_sites = self->tables->sites.num_rows;
+    const tsk_size_t num_samples = self->num_samples;
+    const tsk_size_t max_alleles = self->tables->mutations.num_rows + num_sites;
+    tsk_size_t num_sample_chunks = TSK_BIT_ARRAY_NUM_CHUNKS(num_samples);
+    tsk_size_t *restrict site_offsets = tsk_malloc(num_sites * sizeof(*site_offsets));
+    tsk_size_t *restrict num_alleles = tsk_malloc(num_sites * sizeof(*num_alleles));
+    tsk_bit_array_t *restrict sample_sets
+        = tsk_calloc(num_sample_chunks * state_dim, sizeof(*sample_sets));
+    tsk_bit_array_t *allele_samples
+        = tsk_calloc(max_alleles * num_sample_chunks, sizeof(*allele_samples));
+    tsk_bit_array_t all_samples_bits[num_sample_chunks];
+
+    if (site_offsets == NULL || num_alleles == NULL || sample_sets == NULL
+        || allele_samples == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    // A future improvement could get a union of all sample sets
+    // instead of all samples
+    get_all_samples_bits(all_samples_bits, num_samples, num_sample_chunks);
+
+    tsk_bit_array_t *allele_samples_row;
+    tsk_size_t num_alleles_cumsum = 0;
+    for (tsk_size_t t = 0; t < self->num_trees; t++) {
+        for (tsk_size_t s = 0; s < self->tree_sites_length[t]; s++) {
+            tsk_id_t site_id = self->tree_sites[t][s].id;
+            // Initialize the allele_samples with all samples in the ancestral allele
+            allele_samples_row
+                = GET_2D_ROW(allele_samples, num_sample_chunks, num_alleles_cumsum);
+            tsk_add_bit_arrays(allele_samples_row, all_samples_bits, num_sample_chunks);
+            // Store the allele offset for each site
+            site_offsets[site_id] = num_alleles_cumsum;
+            num_alleles_cumsum += self->site_mutations_length[site_id] + 1;
+        }
+    }
+
+    ret = get_all_mutation_samples(
+        self, num_sample_chunks, site_offsets, num_alleles, &allele_samples);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // Number of pairs w/ replacement (sites)
+    *result_size = (num_sites * (1 + num_sites)) >> (tsk_size_t) 1;
```

Comment:
Is there a good reason to divide by two by bitshifting? It's more readable IMO to write `/ 2`.

Reply:
I didn't realize until recently that you could do truncated division with unsigned integers. It seems that this works with c99 on clang and gcc, I can change this to division so it's more readable. Here's a simple [playground](https://cplayground.com/?p=hamster-porcupine-ape) session showing my findings.

Reply:
+1 on this - any compiler worth it's salt will replace with a bitshift if there's a literal ``integer_value / 2`` in there.